### ISSUE TITLE
[jsonschema] Support for schema.Reference

### ISSUE THREAD

**Issue by smyrman** [rs/rest-layer#36]
This ticket is about adding support for `schema.Reference` to the `schema/encoding/jsonschema` package.

## Dependency on the index
Given  we only want to document a schema for the ID value, we could call `validatorToJSONSchema()` on the related object's ID validator. As you can imagine, this poses a small problem, as we would presumably need access to the index.

One solution would be to pass an (optional) index instance to `jsonschema.Encoder`, either via `NewEncoder()`, or as a later call. E.g.:
```go
func (e *Encoder) SetIndex(index *resourece.Index) {...}
```

If the index is not set, and a schema contain references is encoded to JSON Schema, we could choose to let the validator encoding be a nil operation (but still encode any values from the `schema.Field`), or we could return an error. Note that a JSON Schema without a "type" set, appears to be technically valid according to the [meta spec](http://json-schema.org/schema) (JSON Schema definition of JSON Schema). Hint: there is no "required" propert in the outer JSON object, which means a JSON Schema has no required fields.

## Document embedded objects?
This is possible via [references](http://json-schema.org/latest/json-schema-core.html#anchor30), but probably **not needed** as long as objects can only be embedded for read, and not for write.

If we would want to describe the embedded resource, we would need to use a `oneOf` or `anyOf` clause:
```JavaScript
"anyOf": [
        {/* Schema encoded from the referenced resource ID field's validator */},
        {"$ref": "<uri to definition>", "readOnly": true}
]
```

@rs - as far as I understand [the code](https://github.com/rs/rest-layer/blob/6d2f12fd6e0f4b743ad54be2e43cc887341ff090/rest/util.go#L178), references only support _writing_ IDs, and not JSON versions of the referenced resoure?

**Comment by rs** [comment]
Yes, you can't modify the content of the reference from the parent document in the current state of the code.

**Comment by smyrman** [comment]
> (...) As you can imagine, this poses a small problem, as we would presumably need access to the index.

For JSON Schema used with the Swagger 2.0 specification (a.k.a. OPEN API Specification), we can actually avoid the `resource.Index` dependency completely by using the "$ref" keyword. If we can accept no type to be defined when using JSON Schema without swagger, this solution could be gould enough...

Swagger 2.0 defines all payloads for an API under the "definitions" parameter in the [root schema](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#schema). Therefore, we can refer to the ref validator schema definition as `"myRef": {"$ref": "#/definitions/{TypeName}/id"}`.

Note that {TypeName} is most correctly spelled singular and possibly capitalized, so it should not be taken directly from the reference path. The reference path can also be deep. This must however be solved anyway to enable Swagger 2.0 documentation, so it's not a Reference specific problem.

**Comment by smyrman** [comment]
UPDATE: if #100 is fixed/merged, there is an opportunity to access the Validator of the linked ID we are targeting if the schema is registered with a `resource.Resource` within a `resource.Index` that gets compiled.
